src/script.c

```diff
@@ -0,0 +1,422 @@
+    if (acl_retval != ACL_OK) {
+        addACLLogEntry(c,acl_retval,ACL_LOG_CTX_LUA,acl_errpos,NULL,NULL);
+        switch (acl_retval) {
+        case ACL_DENIED_CMD:
+            *err = sdsnew("The user executing the script can't run this "
+                          "command or subcommand");
+            break;
+        case ACL_DENIED_KEY:
+            *err = sdsnew("The user executing the script can't access "
+                          "at least one of the keys mentioned in the "
+                          "command arguments");
+            break;
+        case ACL_DENIED_CHANNEL:
+            *err = sdsnew("The user executing the script can't publish "
+                          "to the channel mentioned in the command");
+            break;
+        default:
+            *err = sdsnew("The user executing the script is lacking the "
+                          "permissions for the command");
+            break;
+        }
+        return C_ERR;
+    }
+    return C_OK;
+}
+
+static int scriptVerifyWriteCommandAllow(scriptRunCtx *r_ctx, char **err) {
+    if (!(r_ctx->c->cmd->flags & CMD_WRITE)) {
+        return C_OK;
+    }
+
+    if (r_ctx->flags & SCRIPT_READ_ONLY) {
+        /* We know its a write command, on a read only run we do not allow it. */
+        *err = sdsnew("Write commands are not allowed from read-only scripts.");
+        return C_ERR;
+    }
+
+    if ((r_ctx->flags & SCRIPT_RANDOM_DIRTY) && (r_ctx->flags & SCRIPT_EVAL_REPLICATION)) {
+        *err = sdsnew("Write commands not allowed after non deterministic commands. Call redis.replicate_commands() at the start of your script in order to switch to single commands replication mode.");
+        return C_ERR;
+    }
+
+    /* Write commands are forbidden against read-only slaves, or if a
+     * command marked as non-deterministic was already called in the context
+     * of this script. */
+    int deny_write_type = writeCommandsDeniedByDiskError();
+
+    if (server.masterhost && server.repl_slave_ro && r_ctx->original_client->flags != CLIENT_ID_AOF
+        && !(r_ctx->original_client->flags & CLIENT_MASTER))
+    {
+        *err = sdsdup(shared.roslaveerr->ptr);
+        return C_ERR;
+    }
+
+    if (deny_write_type != DISK_ERROR_TYPE_NONE) {
+        if (deny_write_type == DISK_ERROR_TYPE_RDB) {
+            *err = sdsdup(shared.bgsaveerr->ptr);
+        } else {
+            *err = sdsempty();
+            *err = sdscatfmt(*err,
+                    "MISCONF Errors writing to the AOF file: %s\r\n",
+                    strerror(server.aof_last_write_errno));
+        }
+        return C_ERR;
+    }
+
+    return C_OK;
+}
+
+static int scriptVerifyOOM(scriptRunCtx *r_ctx, char **err) {
+    /* If we reached the memory limit configured via maxmemory, commands that
+     * could enlarge the memory usage are not allowed, but only if this is the
+     * first write in the context of this script, otherwise we can't stop
+     * in the middle. */
```

Comment:
We have a known issue where a script can execute a delete first, and then continue to use memory and grow. It seems like we should fix that initially here.

Reply:
Yes we actually have an idea to handle this issue and will share on following PR's/issues. For this PR I wanted to keep the behaviour untouched.

Reply:
Sounds good, we should just make sure it's fixed before 7, since it seems like we'll need some type of breaking change.